### DISCUSSION TITLE
appimagetool - AppStreamData - launchable? Help!!

### DISCUSSION THREAD

**Discussion by kuiperzone** [AppImage/AppImageKit#1244] [Q&A]
I'm having difficulty with an AppStream metadata file. (I'll include my complete appdata.xml file below -- does it look right?)

If I DON'T put in the following line in appdata.xml:

    <launchable type="desktop-id">net.example.AppId.desktop</launchable>

I get the following error: `E: net.example.helloworld:~: desktop-app-launchable-missing`

OK, so I put the above line into the file. But now I get the warning:

`W: net.example.helloworld:~: desktop-file-not-found`

Now, I've discovered I can make this warning go away by putting a copy of the `.desktop` file into `usr/share/applications`. But this feels wrong, as the project already has a copy of the desktop file in the AppDir root. Moreover, I'm not sure the relative paths in the `.desktop` file under `usr/share/applications` would be correct anyway?

How can I solve this error and warning?

***

  <?xml version="1.0" encoding="UTF-8"?>
  <component type="desktop">
      <id>net.example.AppId</id>
      <metadata_license>MIT</metadata_license>
      <project_license>MIT</project_license>
      <content_rating type="oars-1.1" />
  
      <name>Hello World</name>
      <summary>Test application only</summary>
  
      <description>
          <p>This is a longer description. There is a minimum length for this, otherwise the builder tool may
          give you a warning. Replace this with your own here. This is a test.</p>
      </description>
  
      <releases>
          <release version="1.0.0" date="2023-01-01">
              <description><p>This is the latest release.</p></description>
          </release>
      </releases>
  
	  <launchable type="desktop-id">net.example.AppId.desktop</launchable>
  
      <developer_name>Your Name</developer_name>
      <url type="homepage">https://example.net/</url>
  
      <screenshots>
          <screenshot type="default">
              <image>https://via.placeholder.com/800x450.png/0000FF/808080?text=Screenshot+(TBD)</image>
          </screenshot>
      </screenshots>
  
      <provides>
		  <id>net.example.AppId.desktop</id>
	  </provides>
  
  </component>

**Comment by probonopd** [comment]
Yes, the whole AppStream format can be a bit overwhelming, is complex, and is a moving target, making it hard to produce a correct one. Can you generate a working AppStream file using https://docs.appimage.org/packaging-guide/optional/appstream.html#using-the-appstream-generator?

(What happens if you use `helloworld.desktop` as the name of your desktop file in the AppDir, and as the value of `<launchable>` and `<id>`?)

**Reply by kuiperzone** [reply]
Hey - thanks for the reply.

Using the generator, you get the same thing as I included above regarding "launchable". I've just created one, which I'll paste below.

Yes, AppStream data appears an impossible moving target. It didn't used to generate these errors and warnings, and I'm wondering whether the validation tool is using online rules that are always changing.



      <?xml version="1.0" encoding="UTF-8"?>
      <component type="desktop-application">
	      <id>net.example.helloworld</id>
	      <metadata_license>MIT</metadata_license>
	      <project_license>MIT</project_license>
	      <name>HelloWorld</name>
	      <summary>Test application</summary>
	      <description>
		      <p>This is a longer description. There is a minimum length for this, otherwise the builder tool may give you a warning. Replace this with your own here. This is a test.</p>
	      </description>
	      <launchable type="desktop-id">net.example.helloworld.desktop</launchable>
	      <url type="homepage">https://example.net</url>
	      <screenshots>
		      <screenshot type="default">
			      <image>https://via.placeholder.com/800x450.png/0000FF/808080?text=Screenshot+(TBD)</image>
		      </screenshot>
	      </screenshots>
	      <provides>
		      <id>net.example.helloworld.desktop</id>
	      </provides>
      </component>

**Reply by probonopd** [reply]
What happens if you
* Use `helloworld.desktop` instead of `net.example.helloworld.desktop`
* Have corresponding `helloworld.desktop` file(s) in your AppDir

We are not using online rules. But from time to time people are urging us to use newer versions of the validation tool, because the format is ever-changing. Argh.

**Reply by kuiperzone** [reply]
I think saving desktop as myapp.desktop rather than reverse DNS form solves the problem. Thanks.
